### ISSUE TITLE
Reference file overrides need a URI when recorded in the header keyword

### ISSUE THREAD

**Issue by jdavies-st** [STScI-JWST/jwst#1969]
When a Step is run with a reference file override, the full path of that reference file is stored in the header keyword of the output file. This makes the output file not very portable.  Would it be good to have some sort of $OREF equivalent for the JWST pipeline?

This has come up in testing where the "truth" comparison data is produced using an override, and then a test is run and fitsdiff run to compare to the truth data.  If the truth data was produced on a different system from the test data, then the full path to the override reffile will be different.

**Comment by hbushouse** [comment]
The concept of an environment variable like `$oref` only works when reference files are stored in a standard place, e.g. /grp/crds/jwst/. When overrides are used, the ref file is no longer in a standard place and hence an environment variable to represent the path is not feasible. In the past when I've encountered the situation described above, once the test result has been shown to be identical to the "truth" file in every way other than the path to the ref file, then I just update the "truth" file so that it now contains the new path and does not show any more differences in the future.

**Comment by jdavies-st** [comment]
Could we at least strip the absolute path from the reffile override?  That way at least if the reference file lives with the input data it can be portable?

**Comment by hbushouse** [comment]
That seems reasonable. If the path is not the standard crds:// (or whatever the exact syntax is), then strip it before it gets stored in the relevant R_xxxxx keyword. I think that may happen in the CRDS server code somewhere and hence the change would need to be done on that side. @jaytmiller ?

**Comment by jdavies-st** [comment]
I like that.

I think that functionality could be done on the crds_client.py in `stpipe`.

**Comment by jdavies-st** [comment]
And then we can just update the documentation to indicate, "if you are using a reffile override, it should live with the input data so that it is portable."

**Comment by jaytmiller** [comment]
That's the correct syntax.   I think all of the relevant crds:// handling occurs in stpipe.crds_client and stpipe.step,  it's really a JWST-specific syntax.

The relevant code is probably here in stpipe:

    def get_reference_file(self, input_file, reference_file_type):
        """
        Get a reference file from CRDS.

        If the configuration file or commandline parameters override the
        reference file, it will be automatically used when calling this
        function.

        Parameters
        ----------
        input_file : jwst.datamodels.ModelBase instance
            A model of the input file.  Metadata on this input file
            will be used by the CRDS "bestref" algorithm to obtain a
            reference file.

        reference_file_type : string
            The type of reference file to retrieve.  For example, to
            retrieve a flat field reference file, this would be 'flat'.

        Returns
        -------
        reference_file : path of reference file,  a string
        """
        override = self.get_ref_override(reference_file_type)
        if override is not None:
            if override.strip() != "":
                self._reference_files_used.append(
                    (reference_file_type, **abspath(override))**)
                reference_name = override
            else:
                return ""
        else:
            reference_name = crds_client.get_reference_file(
                input_file, reference_file_type)
            if reference_name != "N/A":
                hdr_name = "crds://" + basename(reference_name)
            else:
                hdr_name = "N/A"
            self._reference_files_used.append(
                (reference_file_type, hdr_name))
        return crds_client.check_reference_open(reference_name)

Maybe just tweak the override case to use basename() instead of abspath()?  (I'm not being sarcastic,  don't know implications of it.)

**Comment by jdavies-st** [comment]
Agree @jaytmiller.  The other thing to verify would be that it works properly in the pipeline environment - i.e. if my data and the override reffile is in an input directory not in my current working directory which is where the output gets sent, then it needs to be able to find the override reffile based on the input data abspath.

**Comment by philhodge** [comment]
I don't think the documentation should say that the overriding reference file has to be in the same directory as the data.  We ought to be able to include a path to another directory, and the file should be read from that directory.  It's fine with me if the file name without the path is recorded in the output header, though.

**Comment by hbushouse** [comment]
Agreed. The user should still be able to include a path in the path to the file specified via the `override_xxxx` argument, but the path will be stripped before the name of that file is stored in an R_xxxxx keyword (which is informational only; it is not the input to the step).

**Comment by hbushouse** [comment]
@jdavies-st Are you suggesting that in a case like:
```
$ strun <step.cfg> </my/file/path/file.fits> --override_flat='my_flat.fits'
```
it should by default look for `my_flat.fits` in `/my/file/path/` (the same path as the input data)? I'm not sure I agree with that. If the user expects the override file to come from some path other than the working directory, they should include that path explicitly in the `override_xxxxx` argument (IMHO).

**Comment by jdavies-st** [comment]
I agree. That seems reasonable.

So only for the case when a file is listed in the keyword of a parent file, should the assumption be made that it resides with the parent file.  In that case, we want the full base path appended the file when opened.

I guess this issue can be tracked via #1981 as it is the same general issue.

**Comment by hbushouse** [comment]
Fixed in #1995
